crates/sui-json-rpc/src/indexer_api.rs

```diff
@@ -371,26 +370,103 @@ impl<R: ReadApiServer> IndexerApiServer for IndexerApi<R> {
-                Error::UnexpectedError(format!(
-                    "Failed to parse NameService Domain with error: {:?}",
-                    e
-                ))
-            })?;
+            let requested_domain = name.parse::<Domain>().map_err(Error::from)?;
+
+            // Get the SLD name, if we're operating on a subdomain, or use the requested as is.
+            let sld = if requested_domain.is_subdomain() {
+                requested_domain.sld()
+            } else {
+                requested_domain.clone()
+            };
 
-            let record_id = self.name_service_config.record_field_id(&domain);
+            let sld_record_id = self.name_service_config.record_field_id(&sld, None);
 
-            let field_record_object = match self.state.get_object(&record_id).await? {
-                Some(o) => o,
-                None => return Ok(None),
+            let Some(field_record_object) = self.state.get_object(&sld_record_id).await? else {
+                return Ok(None);
             };
 
-            let record = field_record_object
-                .to_rust::<Field<Domain, NameRecord>>()
-                .ok_or_else(|| Error::UnexpectedError(format!("Malformed Object {record_id}")))?
-                .value;
+            let sld_record = NameRecord::try_from(field_record_object)?;
+
+            // Check if the record is expired.
+            if sld_record.is_expired() {
+                return Err(Error::from(NameServiceError::NameExpired));
+            }
+
+            // For SLD, we can resolve here.
+            if !requested_domain.is_subdomain() {
+                return Ok(sld_record.target_address);
+            }
+            // == Doing a subdomain lookup from this point forward ==
+
+            // Find the namespace from the name record's metadata.
+            let namespace_table_id = sld_record.namespace_table_id();
+
+            if namespace_table_id.is_none() {
+                return Ok(None);
+            }
+
+            // Look-up the subdomain record object.
+            let sub_record_object_id = self
+                .name_service_config
+                .record_field_id(&requested_domain, namespace_table_id);
+
+            let subdomain_parent = requested_domain.parent();
+
+            let sub_parent_record_object_id = self
+                .name_service_config
+                .record_field_id(&subdomain_parent, namespace_table_id);
+
+            // Flag if the subdomain's parent is also a subdomain.
+            let parent_is_subdomain = requested_domain.depth() > 3;
+
+            // Gather all requests to execute concurrently.
+            let mut requests = vec![self.state.get_object(&sub_record_object_id)];
+
+            // We only query the parent if we're looking at a subdomain whose parent is also a subdomain.
+            // Subdomains on the first level (depth === 3) have the SLD as their parent, and we already have the SLD fetched & validated.
```

Comment:
```suggestion
            // Subdomains on the first level (depth == 3) have the SLD as their parent, and we already have the SLD fetched & validated.
```

No JavaScript here ;)

Reply:
you caught me.